File: Ioana_Hisham/src/main/java/ql/visitors/StatementVisitor.java
package ql.visitors;

import ql.ast.statements.CalculableQuestion;
import ql.ast.statements.IfThen;
import ql.ast.statements.IfThenElse;
import ql.ast.statements.Question;

public interface StatementVisitor<T> {
    T visit(CalculableQuestion calculableQuestion);
    T visit(IfThen ifThen);
    T visit(IfThenElse ifThenElse);
    T visit(Question question);
}

File: niro/src/main/scala/nl/uva/se/sc/niro/gui/StatementVisitor.scala
package nl.uva.se.sc.niro.gui

import javafx.scene.layout._
import nl.uva.se.sc.niro.model.Expressions.Expression
import nl.uva.se.sc.niro.model._

object StatementVisitor {

  def visit(grid: GridPane, statements: Seq[Statement], symbolTable: Map[String, Expression]): Unit = {
    grid.setHgap(10)

    statements.zipWithIndex foreach {
      case (statement, row) => statement match {
        case question: Question => {
          RowBuilder.buildDoubleColumnRow(grid, row)
          WidgetFactory.makeWidgets(question, symbolTable).zipWithIndex foreach {
            case (widget, column) => grid.add(widget, column, row)
          }
        }
        case conditional: Conditional => {
          val innerGrid = RowBuilder.buildSingleColumnRow(grid, row)
          // FIXME and make dynamic!
//          innerGrid.setVisible(conditional.predicate.value)
          visit(innerGrid, conditional.thenStatements, symbolTable)
        }
        case ErrorStatement() => ()
      }
    }
  }

}
